File: src/main/java/bad/robot/radiate/teamcity/UnexpectedContentType.java
package bad.robot.radiate.teamcity;

import bad.robot.http.Header;
import bad.robot.http.HttpResponse;
import com.googlecode.totallylazy.Predicate;
import org.apache.commons.lang3.StringUtils;

import static com.googlecode.totallylazy.Sequences.sequence;
import static java.lang.String.format;

@Deprecated
class UnexpectedContentType extends TeamCityException {

    public UnexpectedContentType(HttpResponse response) {
        super(format("Unexpected response format '%s' (%s %s) from %s%n%s", sequence(response.getHeaders()).find(contentType()).get().value(), response.getStatusCode(), response.getStatusMessage(), response.getOriginatingUri(), abbreviated(response)));
    }

    private static String abbreviated(HttpResponse response) {
        String body = response.getContent().asString();
        return StringUtils.abbreviate(body, 400);
    }

    private static Predicate<Header> contentType() {
        return header -> header.name().equalsIgnoreCase("content-type");
    }

}

File: src/main/scala/bad/robot/radiate/teamcity/UnexpectedContentType.scala
package bad.robot.radiate.teamcity

import bad.robot.http.HeaderPair._
import bad.robot.http.{Header, HttpResponse}
import bad.robot.radiate.teamcity.UnexpectedContentType._
import org.apache.commons.lang3.StringUtils

object UnexpectedContentType {
  private def abbreviated(response: HttpResponse): String = {
    val body = response.getContent.asString
    StringUtils.abbreviate(body, 400)
  }

  private def contentType(response: HttpResponse): Header = {
    import scala.collection.JavaConverters._
    response.getHeaders.asScala.find(header => header.name().equalsIgnoreCase("content-type")).getOrElse(header("content-type", ""))
  }
}

class UnexpectedContentType(response: HttpResponse) extends TeamCityException(
  s"Unexpected response format '${contentType(response)}' (${response.getStatusCode} ${response.getStatusMessage}) from ${response.getOriginatingUri}\n${abbreviated(response)}"
)